enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+additional workers added to them. The absence of multiple control-plane nodes
+means that the number of workers/pods that can be supported on such a topology is
+even more limited than in a regular 3 control-plane node cluster. Relevant
+documentation may have to point out that the pod limits (or other control-plane
+related limits) that apply to a single control-plane node cluster with no
+workers also apply globally across all additional added workers. If a large
+amount of workers/pods is desired the user would have to re-install the cluster
+as a regular multi control-plane node cluster.
+
+- Deal with clusters that have less nodes than they had during day-1
+installation. The rationale for this being a non-goal is that it seems that
+removing the "initial" nodes of a cluster causes a lot of issues even in a
+regular, multi-node installation. So dealing with similar issues in single
+control-plane node clusters is also out of scope for this enhancement. For
+example, installing a cluster with 3 control-plane nodes and 3 workers results
+in unschedulable control-plane nodes. Scaling down the worker machine-sets from
+3 to 0 post-installation does not magically make the control-plane nodes
+schedulable, and as a result a lot of infrastructure workloads fail to schedule
+and operators degrade. User intervention is required to fix this.
+
+- Deal with expansion of the single-node control-plane by adding more
+control-plane nodes
+
+- Deal with expansion of clusters that have been installed before the
+implementation of this enhancement (if possible, their expansion may be
+addressed by an upgrade followed with documentation outlining the steps that
+the user has to take to enable expansion, e.g. patch the default
+`IngressController` to have node selector targetting the master pool rather
+than the worker pool).
+
+- Adjust the baremetal platform to support single control-plane node
+installations. The baremetal platform solves the "floating ingress" issue by
+using virtual IP addresses/keepalived.
+
+- Deal with users who want their ingress traffic to not go through the single
+control-plane node.
+
+- Deal with non-cloud, non-"none" platforms such as baremetal, vSphere, etc.
+
+## Proposal
+
+- Create a new `IngressPlacement` API parameter.
+
+- Make sure worker node ignition files are generated even in bootstrap-in-place
+single control-plane node "none"-platform installation.
+
+### User Stories
+
+- As an OpenShift cluster administrator, I want to install a single control-plane
+node cluster with additional worker nodes, so that my cluster will be able to
+handle larger computation demands.
+
+- As an OpenShift cluster administrator, I want to add worker nodes to my
+existing single control-plane node cluster, so that it'll be able to meet
+growing computation demands.
+
+### API Extensions
+
+Introduce a new field in the Ingress config CR
+(`config.openshift.io/v1/ingresses`) called `IngressPlacement`.
```

Comment:
Is it possible we could flesh out the API here in Go format to match what will be added to the API, eg

```
type IngressSpec struct {
  // Existing fields omitted

  // ingressPlacement ...
  // +kubebuilder:validation:Enum:="ControlPlane";"Worker"
  // +kubebuidler:default:="Worker"
  // +optional
  IngressPlacement string `json:"ingressPlacement"`
}
```

It would also be good to think about a few things here:
- What does the field do and how will it be used by users
- Given the existing possible values mentioned below, do we expect that there may ever be any other configuration to the placement? Eg if this were a struct it could have a field for `role` which would select on the role of the Machine, and then maybe some other fields in the future as needed?
- Given this field is on the `Ingress` object, does it need to be `IngressPlacement` or should it just be `Placement`?
- We are making an effort to ensure appropriate kubebuilder validation is present on all fields going into the API, so it would be good to represent that here, is it optional/required for example, if it's an Enum, then add an Enum tag or make it a typed string which will generate the same OpenAPI validation

Reply:
Yeah I started work on doing something similar in the openshift/api repo. Since your comments lowered my confidence in this solution, I'll wait for us to agree before making this change to the enhancement